go-runtime/ftl/option.go

```diff
@@ -199,3 +198,23 @@ func (o Option[T]) GoString() string {
 	}
 	return fmt.Sprintf("None[%T]()", o.value)
 }
+
+func (o Option[T]) Marshal(w *bytes.Buffer, encode func(v reflect.Value, w *bytes.Buffer) error) error {
+	if o.ok {
+		return encode(reflect.ValueOf(&o.value).Elem(), w)
+	}
+	w.WriteString("null")
+	return nil
+}
+
+func (o *Option[T]) Unmarshal(d *json.Decoder, isNull bool, decode func(d *json.Decoder, v reflect.Value) error) error {
+	if isNull {
```

Comment:
I still don't quite understand why we're replacing `json.Unmarshaller` with custom unmarshallers, can you explain it?

Reply:
because we need to pass `context.Context` into our `encoding.Marshal/Unmarshal`

Reply:
this change will be followed by one that pipes `ctx` through, and we'll use the type registry stored in the context to manage sum type encoding

the `MarshalJSON/UnmarshalJSON` implementations would necessarily need to call into `encoding.Marshal/Unmarshal` to serialize the underlying optional value but we wouldn't be able to pass context from there

Reply:
But the question is why we're not using the standard JSON decoder/encoder directly?

Reply:
> this change will be followed by one that pipes ctx through, and we'll use the type registry stored in the context to manage sum type encoding



Aaah, understood, that was the piece of the puzzle I was missing.